### PR TITLE
Add more stats to nuts and some refactor

### PR DESCRIPTION
**Pull request by aseyboldt**
I added a couple of stats to the output of nuts and refactored the algorithm a little along the way. I also found two issues in the implementation:
- The mean acceptance probability along a trajectory is using only the last half of the binary tree. This is the way the algorithm in the paper does it, but I can't see a good reason for doing this. If I understand the code in stan right, it uses the whole tree. This would be really easy to fix: just change the `=`in line 212 and 214 to a `+=`.
- The stopping criterion differs in this implementation from that in the paper. See line 251. I don't know which one is right. The lines that are commented out is the version from the paper.

I still want to write a couple more tests and update the stats notebook (coming soon).

### CODE REVIEW

File: pymc3/step_methods/hmc/nuts.py [hunk 1]
```diff
@@ -155,35 +137,135 @@ def competence(var):
...
+Edge = namedtuple("Edge", 'q, p, q_grad, energy')
+
+# A proposal for the next position
+Proposal = namedtuple("Proposal", "q, energy, p_accept")
+
+# A subtree of the binary tree build by nuts.
+Subtree = namedtuple("Subtree", "left, right, proposal, depth, size, accept_sum, n_proposals")
+
+
+class Tree:
```
Comment (twiecki) [inline pymc3/step_methods/hmc/nuts.py]: `(object)`

**Comment by twiecki** [comment]
Interesting. Assigning @ColCarroll who did the latest refactoring (feel free to unassign, Col).

As to the implementaiton differences you found. Stan also switched from a Slice sampler to a Multinomial, could that be related? In any case, we should make the same change here eventually.

As to adding more stats, it would be great if we could generate these energy plots:
image

More detail: https://arxiv.org/pdf/1701.02434v1.pdf

**Comment by aseyboldt** [comment]
I don't think the implementation issues are related to the slice/multinomial difference. The first is unrelated to sampling but only about step size adaptation, the second is a difference between the implementation and the algorithm in the paper.

The plots look easy enough, I'll put that in the notebook.

Somehow I can't find a paper about the multinomial sampling. Do you know where to find that?

File: pymc3/step_methods/hmc/nuts.py (hunk 1 above)
Comment (aseyboldt) [reply pymc3/step_methods/hmc/nuts.py]: I hate python2 compatibility. ;-)

**Comment by twiecki** [comment]
It's mentioned in passing in the exhaustive HMC paper by Betancourt.

**Comment by ColCarroll** [comment]
This looks really cool, and much more readable than in the past.  I'd noticed the mean acceptance probability as well, but deferred to the paper -- we should definitely take a closer look at that.

I'm a little busy for the next ~24hrs, but can do a proper review later this week.  I think the only thing I'm worried about is whether this will make it harder to convert this to using pure theano (though it looks like maybe this wouldn't hurt that?).

**Comment by aseyboldt** [comment]
@ColCarroll take your time, no rush :-)

I also thought a bit about using theano for it all, but I have to admit I'm a bit at a loss about that. I guess the first step would be to convert it to an iterative algorithm, maybe it's easier to see then. It might also help to use shared variables for all the q, p and q_grad values, that would at least save all the transfers between gpu and cpu when this is running on a gpu (those were about 20% of the runtime in a simple experiment, but this probably depends a lot on the model and hardware).

**Comment by twiecki** [comment]
Using theano shared is a great idea. Didn't realize that about the GPU.

**Comment by twiecki** [comment]
That actually makes me wonder -- perhaps we should replace all RV values with theano shared's on which the step methods call `.set_value()`. That's probably faster, could be cleaner, and work better with the GPU.

**Comment by aseyboldt** [comment]
I removed the estimator test in test_samples and wrote new tests, that should be more robust. Is the posterior in the original test normal at all? As to what posteriors to test, I'm open to ideas. Something with correlations and something with a bit of a funnel would be great.

**Comment by twiecki** [comment]
Tests look great, one idea would be to also test Rhat, although that's sort of implied by the other tests.

I'm eager to get this in, especially because of the bug fix. @aseyboldt Is this ready from your end?

**Comment by aseyboldt** [comment]
Rhat is tested already. :-)
From my end this is ready to merge. A few comments:
- The question about how to compute the acceptance probability is still open. I don't think that that is a blocker for merging this.
- I included a test that takes a long time to complete, but is not executed by default. It can be run using `nosetests -a extra`. We might also want to write more of those. They should probably at least be run before releases.

**Comment by twiecki** [comment]
@aseyboldt Oh, didn't see that -- great.

The `-a extra` approach seems like an elegant solution for longer running tests.

Will let @ColCarroll give the final go.

**Comment by twiecki** [comment]
@aseyboldt Also, are we any closer to that energy plot above?

**Comment by aseyboldt** [comment]
Sorry, forgot about that. It's in the notebook now.

**Comment by twiecki** [comment]
@aseyboldt Fantastic, appreciate it. btw are you on gitter? Tried to contact you there.

**Comment by aseyboldt** [comment]
I was just working on a project using the new sampler stats and managed to get confused about the meaning of one of the names. As they are part of the public api, do you want to double check them too? I now think we should rename `energy_change` to `energy_error` and `max_energy_change` to `max_energy_error`. Here are the statistics and some description from the notebook:

- mean_tree_accept: The mean acceptance probability for the tree that generated this sample. The mean of these values across all samples but the burn-in should be approximately target_accept (the default for this is 0.8).
- diverging: Whether the trajectory for this sample diverged. If there are many diverging samples, this usually indicates that a region of the posterior has high curvature. Reparametrization can often help, but you can also try to increase target_accept to something like 0.9 or 0.95.
- energy: The energy at the point in phase-space where the sample was accepted. This can be used to identify posteriors with problematically long tails. See below for an example.
- energy_change: The difference in energy between the start and the end of the trajectory. For a perfect integrator this would always be zero.
- max_energy_change: The maximum difference in energy along the whole trajectory.
- depth: The depth of the tree that was used to generate this sample
- tree_size: The number of leafs of the sampling tree, when the sample was accepted. This is usually a bit less than $2 ^ \text{depth}$. If the tree size is large, the sampler is using a lot of leapfrog steps to find the next sample. This can for example happen if there are strong correlations in the posterior, if the posterior has long tails, if there are regions of high curvature ("funnels"), or if the variance estimates in the mass matrix are inaccurate. Reparametrisation of the model or estimating the posterior variances from past samples might help.
- tune: This is True, if step size adaptation was turned on when this sample was generated.
- step_size: The step size used for this sample.
- step_size_bar: The current best known step-size. After the tuning samples, the step size is set to this value. This should converge during tuning.

**Comment by fonnesbeck** [comment]
Sorry I'm getting to this late. It would be great if `step_size_bar` plotted separate lines (or returned a faceted plot). This should be easy to do if you use call the trace's `get_values` method with `combine=False`, which will return them unconcatenated.

**Comment by aseyboldt** [comment]
Do you mean the plot in the notebook? `step_size_bar` does not plot anything on its own. I'll do that tomorrow.

**Comment by twiecki** [comment]
@aseyboldt I read through your descriptions and energy_error makes sense. The other descriptions are helpful too.

**Comment by aseyboldt** [comment]
I renamed `energy_change` and `max_energy_change` and made a faceted plot of one of the sampler stats in the notebook. I still left in one where the values are concatenated, because I think people might stumble over that behaviour anyway. I don't think it's easy to change, or the interface for the stats would have to be very different from the interface for the normal variables.

**Comment by twiecki** [comment]
Yeah, that's fine. I think we should merge, this is very worthwhile.

**Comment by twiecki** [comment]
👍

**Pull request merged**